### PR TITLE
feat(@aws-amplify/datastore): ObserveQuery performance and type enhancements

### PR DESCRIPTION
**Pull request by nickarocho**
#### Description of changes

- constrains ObserveQuery options with an explicit, unique Type
- adds a time interval/limit Promise race to return data when Model datasets are very large (i.e. have lots of fields in the Schema)

I'll be re-working the sample app/integ tests to utilize this data set to account for worst-case scenarios

#### Description of how you validated changes
`yarn test`, manual testing - adding integ test and reworking sample app next

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/src/datastore/datastore.ts [hunk 1]
```diff
@@ -1190,21 +1209,49 @@ class DataStore {
...
 						}
+
+						// kicks off every subsequent race as results sync down
+						if (!raceInFlight) racePromises();
 					});
 
-					// will return any locally-available items in the first snapshot
-					generateAndEmitSnapshot();
+					// kicks off the first race to return any initial results
+					if (!raceInFlight) racePromises();
```
Comment (iartemiev) [inline packages/datastore/src/datastore/datastore.ts:1224]: Why not just return all the results from the local store right away?

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -1190,21 +1209,49 @@ class DataStore {
...
 					observer.error(err);
 				}
 			})();
 
+			const racePromises = async (): Promise<any> => {
+				raceInFlight = true;
+				startTimer();
+
+				try {
+					await Promise.race([timerPromise, limitPromise]);
```
Comment (iartemiev) [inline packages/datastore/src/datastore/datastore.ts:1235]: I know this already works as expected, but I think technically, we would want `limitPromise.promise` here, i.e. the promise itself not just the object that contains that promise as one of its properties

File: packages/datastore/src/datastore/datastore.ts [hunk 3]
```diff
@@ -1152,23 +1153,41 @@ class DataStore {
...
 			const items = new Map<string, T>();
 			const itemsChanged = new Map<string, T>();
 			let deletedItemIds: string[] = [];
 			let handle: ZenObservable.Subscription;
 
+			// a mechanism to return data after X amount of seconds OR after the
+			// "limit" (itemsChanged >= this.syncPageSize) has been reached, whichever comes first
+			const TIMER_INTERVAL = 2000;
+			let timer: NodeJS.Timer;
+			let timerPromise: Promise<any>;
```
Comment (iartemiev) [inline packages/datastore/src/datastore/datastore.ts:1173]: currently, this is `Promise<string>` but since we don't use the resolved value anywhere, we can change line 1250 to `resolve()` and make this `Promise<void>`

File: packages/datastore/src/datastore/datastore.ts [hunk 4]
```diff
@@ -1152,23 +1153,41 @@ class DataStore {
...
+			function LimitPromise(): void {
```
Comment (iartemiev) [inline packages/datastore/src/datastore/datastore.ts:1174]: Lets rename this to something more generic and move it to util. Can be e.g. `DeferredPromise`, `OpenPromise`, etc.

File: packages/datastore/src/datastore/datastore.ts (hunk 1 above)
Comment (nickarocho) [reply packages/datastore/src/datastore/datastore.ts]: Yup, another good call.

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (nickarocho) [reply packages/datastore/src/datastore/datastore.ts]: True, kind of interesting how the `Promise.race` worked with just the object and not the promise itself 🤷🏻‍♂️

**Review by nickarocho — COMMENTED** [review]
Thanks for the feedback, Ivan! Helpful and thorough as always. I've made some adjustments if you want to give it another look.

File: packages/datastore/src/datastore/datastore.ts [hunk 5]
```diff
@@ -1190,21 +1203,48 @@ class DataStore {
...
-							generateAndEmitSnapshot();
+						const limit =
+							itemsChanged.size - deletedItemIds.length >= this.syncPageSize;
+
+						if (limit || isSynced) {
+							limitPromise.resolve();
 						}
+
+						// kicks off every subsequent race as results sync down
+						if (!raceInFlight) racePromises();
```
Comment (svidgen) [inline packages/datastore/src/datastore/datastore.ts:1214]: Trending towards spaghetti, IMO. And I don't see any unit test changes to accompany the change in behavior. Is there an opportunity to factor out some of the time-limited-snapshot mechanism (_buffer_?) into its own little testable class? (`MessageBuffer`? `SubscriptionBuffer`?) I could see that exercise simplifying both the buffering logic and the `observeQuery` method. (I could also see it being a fool's errand!)

What do you think? Opportunity or fool's errand?

Comment (svidgen) [reply packages/datastore/src/datastore/datastore.ts:1214]: This comment is intended to apply generally to the buffering mechanism, not just these LOC, FWIW! I just didn't want to select every line in the diff to make this comment!

File: packages/datastore/src/util.ts [hunk 6]
```diff
@@ -681,3 +682,71 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+		self.resolve = resolve;
+		self.reject = reject;
+	});
+}
+
+export class SubscriptionBuffer {
+	private limitPromise = new DeferredPromise();
+	private timerPromise: Promise<string>;
+	private TIMER_INTERVAL = 2000;
+	private timer: NodeJS.Timer;
```
Comment (jamesaucode) [inline packages/datastore/src/util.ts:698]: It feels weird to have a `NodeJS` type here. Do you think it's better to use `ReturnType<typeof setTimeout>` instead? 😄 This way we can just let Typescript infer the type for us.
```suggestion
	private timer: ReturnType<typeof setTimeout>;
```

File: packages/datastore/src/util.ts [hunk 7]
```diff
@@ -681,3 +682,71 @@ export const isAWSIPAddress = (val: string): boolean => {
...
 	);
 };
+
+export function DeferredPromise(): void {
+	const self = this;
+	this.promise = new Promise((resolve, reject) => {
+		self.resolve = resolve;
+		self.reject = reject;
+	});
+}
```
Comment (iartemiev) [inline packages/datastore/src/util.ts:692]: We can re-write this into a class like this:

```suggestion
class DeferredPromise {
	public promise: Promise<void>;
	public resolve: () => void;
	public reject: () => void;

	constructor() {
		const self = this;
		this.promise = new Promise((resolve, reject) => {
			self.resolve = resolve;
			self.reject = reject;
		});
	}
}
```

File: packages/datastore/__tests__/utils.test.ts [hunk 8]
```diff
@@ -456,4 +457,57 @@ _deleted`;
...
+			});
+
+			expect(buffer.getIsRaceInFlight()).toBe(false);
+
+			const proto = Object.getPrototypeOf(buffer);
+			const spyOnRace = jest.spyOn(proto, 'racePromises');
+
+			buffer.start();
+			expect(buffer.getIsRaceInFlight()).toBe(true);
+			expect(spyOnRace).toBeCalledTimes(1);
```
Comment (iartemiev) [inline packages/datastore/__tests__/utils.test.ts:482]: We should expect `bufferCallback` to be called as well

File: packages/datastore/__tests__/utils.test.ts [hunk 9]
```diff
@@ -456,4 +457,57 @@ _deleted`;
...
+			buffer.resolveBasePromise();
+			buffer.close();
+		});
+
```
Comment (iartemiev) [inline packages/datastore/__tests__/utils.test.ts:486]: I think it would be nice if we could test both happy paths, i.e. "limit wins" as well as "timer wins". We can discuss how to accomplish that

File: packages/datastore/src/util.ts [hunk 10]
```diff
@@ -681,3 +682,71 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+	private callback = () => {};
+	private errorHandler: (error: string) => void;
+	private defaultErrorHandler = (msg = 'SubscriptionBuffer error'): void => {
+		throw new Error(msg);
+	};
+
+	constructor(options: SubscriptionBufferOptions) {
+		this.callback = options.callback;
+		this.errorHandler = options.errorHandler || this.defaultErrorHandler;
+		this.TIMER_INTERVAL = options.maxInterval;
```
Comment (iartemiev) [inline packages/datastore/src/util.ts:709]: Same issue as we were seeing previously with the errorHandler. `options.maxInterval` will always overwrite `TIMER_INTERVAL` here. If we want it to default to 2000, we'd want to do:
```suggestion
		this.timerInterval = options.maxInterval || 2000;
```
Also, no need to make that property all caps, since it's configurable

**Review by iartemiev — CHANGES_REQUESTED** [review]
Looks good! Just some minor changes requested

File: packages/datastore/__tests__/utils.test.ts (hunk 9 above)
Comment (nickarocho) [reply packages/datastore/__tests__/utils.test.ts:493]: Thank you for helping me think about this and implement paths for each!

**Review by iartemiev — APPROVED** [review]
LGTM! 🌮

File: packages/datastore/src/util.ts [hunk 11]
```diff
@@ -681,3 +682,79 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+		this.promise = new Promise(
+			(resolve: (value: string | PromiseLike<string>) => void, reject) => {
+				self.resolve = resolve;
+				self.reject = reject;
+			}
+		);
+	}
+}
+
+export class SubscriptionBuffer {
```
Comment (svidgen) [inline packages/datastore/src/util.ts:701]: At a glance, this looks like it successfully extracts your limits race into something testable. That's a win in my book!

That said, this didn't end up being a _buffer_, really. If we care about naming, I think we need another one for this class. If I'm thinking about how we might re-use this elsewhere, I think sticking with some "racing" or "time limiting" terminology is more appropriate. IMO, it's more of a `TimeLimitForPromise` or `PromiseTimeLimit` &mdash; _or something like that_.

File: packages/datastore/src/util.ts [hunk 12]
```diff
@@ -681,3 +682,79 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+
+	public start(): void {
+		if (!this.raceInFlight) this.racePromises();
+	}
+
+	public close(): void {
+		clearTimeout(this.timer);
+	}
+
+	public resolveBasePromise(): void {
```
Comment (svidgen) [inline packages/datastore/src/util.ts:757]: There's no other resolve-oriented method on this. We could probably shorten this to `resolve`.

File: packages/datastore/src/util.ts [hunk 13]
```diff
@@ -681,3 +682,79 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+	constructor(options: SubscriptionBufferOptions) {
+		this.callback = options.callback;
+		this.errorHandler = options.errorHandler || this.defaultErrorHandler;
+		this.maxInterval = options.maxInterval || 2000;
+	}
+
+	private startTimer(): void {
+		this.timerPromise = new Promise((resolve, reject) => {
+			this.timer = setTimeout(() => {
+				resolve('TIMER');
```
Comment (svidgen) [inline packages/datastore/src/util.ts:722]: If invoking code is expected to use this, I'd strongly suggest using an exported `const` or `enum`.

File: packages/datastore/src/util.ts [hunk 14]
```diff
@@ -681,3 +682,79 @@ export const isAWSIPAddress = (val: string): boolean => {
...
+	public start(): void {
+		if (!this.raceInFlight) this.racePromises();
+	}
+
+	public close(): void {
+		clearTimeout(this.timer);
+	}
+
+	public resolveBasePromise(): void {
+		this.limitPromise.resolve('LIMIT');
```
Comment (svidgen) [inline packages/datastore/src/util.ts:758]: Ditto `const`/`enum` comment.

**Review by svidgen — COMMENTED** [review]
Not blocking. But a few things to consider.

File: packages/datastore/src/util.ts (hunk 11 above)
Comment (nickarocho) [reply packages/datastore/src/util.ts]: Renamed to `DeferredCallbackResolver` via @iartemiev 's input.

**Review by svidgen — APPROVED** [review]

**Comment by OskarD** [comment]
My generated models stopped working after upgrading to this version in aws-amplify.

When I tried to query the model, I got this error: `DataStore - The model is not an instance of a PersistentModelConstructor`

I also got this error everywhere in my application:
```
Call Stack
DataStore.<anonymous>
file:///Users/oskar/code/minna/minna-webapp/node_modules/aws-amplify/node_modules/@aws-amplify/datastore/lib/datastore/datastore.js (589:86)
step
file:///Users/oskar/code/minna/minna-webapp/node_modules/aws-amplify/node_modules/@aws-amplify/datastore/lib/datastore/datastore.js (44:23)
Object.next
file:///Users/oskar/code/minna/minna-webapp/node_modules/aws-amplify/node_modules/@aws-amplify/datastore/lib/datastore/datastore.js (25:53)
fulfilled
file:///Users/oskar/code/minna/minna-webapp/node_modules/aws-amplify/node_modules/@aws-amplify/datastore/lib/datastore/datastore.js (16:58)
```

Downgrading to `aws-amplify@4.3.5` and `@aws-amplify/datastore@352` fixed this problem

**Comment by iartemiev** [comment]
@OskarD if you have **both** `aws-amplify` and `@aws-amplify/datastore` installed, that's likely what is causing the issue. Installing `aws-amplify` includes all of the packages already, so if you also install them individually, that can cause unexpected behavior. Please uninstall `@aws-amplify/datastore` and do a fresh install of your dependencies
```
npm uninstall @aws-amplify/datastore
rm -rf node_modules package-lock.json
npm install
```
If you're still getting the error after that, please open a new GitHub issue, so that we can track it and work on a fix.

**Comment by OskarD** [comment]
Thanks, it seems to be working.

I have just followed AWS guides, how do I know which `@aws-amplify/X` to require and which are included? For instance, `@aws-amplify/ui-components` and `@aws-amplify/ui-react` both need to be required in order to use their exports, right? Or is there a better way to do it?
